js/private/js_run_devserver.js

```diff
@@ -0,0 +1,209 @@
+const path = require('path')
+const perf_hooks = require('perf_hooks')
+const fs = require('fs')
+const os = require('os')
+const child_process = require('child_process')
+
+// Globals
+const RUNFILES_ROOT = path.join(
+    process.env.JS_BINARY__RUNFILES,
+    process.env.JS_BINARY__WORKSPACE
+)
+const synced = new Map()
+const mkdirs = new Set()
+
+// Returns true if a file's stats indicate it is an executable file.
+function isExecutable(stats) {
+    return (
+        !!(stats.mode & fs.constants.S_IXUSR) ||
+        !!(stats.mode & fs.constants.S_IXGRP) ||
+        !!(stats.mode & fs.constants.S_IXOTH)
+    )
+}
+
+// Ensure that a directory exists. If it has not been previously created or does not exist then it
+// creates the directory, first recursively ensuring that its parent directory exists.
+// Intentionally synchronous to avoid race conditions between sync promises
+function mkdirpSync(p) {
+    if (!p) {
+        return
+    }
+    if (mkdirs.has(p)) {
+        return
+    }
+    if (!fs.existsSync(p)) {
+        mkdirpSync(path.dirname(p))
+        fs.mkdirSync(p)
+    }
+    mkdirs.add(p)
+}
+
+// Recursively copies a file, symlink or directory to a destination. If the file has been previously
+// synced it is only re-copied if the file's last modified time has changed since the last time that
+// file was copied. Symlinks are not copied but instead a symlink is created under the destination
+// pointing to the source symlink.
+async function syncRecursive(src, dst) {
+    try {
+        const lstat = await fs.promises.lstat(src)
+        const last = synced.get(src)
+        if (!lstat.isDirectory() && last && lstat.mtimeMs == last) {
+            // this file is already up-to-date
+            return 0
+        }
+        const exists = synced.has(src) || fs.existsSync(dst)
+        synced.set(src, lstat.mtimeMs)
+        if (lstat.isSymbolicLink()) {
+            if (process.env.JS_BINARY__LOG_DEBUG) {
+                console.error(
+                    `Syncing symlink ${src.slice(RUNFILES_ROOT.length + 1)}`
+                )
+            }
+            if (exists) {
+                await fs.promises.unlink(dst)
+            } else {
+                mkdirpSync(path.dirname(dst))
+            }
+            await fs.promises.symlink(src, dst)
+            return 1
+        } else if (lstat.isDirectory()) {
+            const contents = await fs.promises.readdir(src)
+            let totalSynced = 0
+            await Promise.all(
+                contents.map(async (entry) => {
```

Comment:
If the src directory is empty and there's no recursion then I don't think it will be created at all?

Reply:
The returned count will also be 0 in that case. Would that case never actually occur?

Reply:
Might if there is a tree artifact that contains an empty directory

Reply:
Fixed